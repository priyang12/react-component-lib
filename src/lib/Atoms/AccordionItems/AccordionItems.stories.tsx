import {
   AccordionButton,
   AccordionContent,
   AccordionIcon,
} from './AccordionItems';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import { useAccordion } from '../../../Hooks/useAccordion';
import { AccordionContext } from '../../Module/Accordion/Accordion';
export default {
   title: 'Atoms/AccordionItems',
   component: AccordionButton,
   decorators: [story => <div className="container">{story()}</div>],
} as ComponentMeta<typeof AccordionButton>;

export const Template: ComponentStory<typeof AccordionButton> = args => {
   const { Indexes, handleItemClick } = useAccordion({
      initialOpenIndexes: [],
   });

   return (
      <AccordionContext.Provider
         value={{
            indexes: Indexes,
            handleItemClick,
         }}
      >
         <AccordionButton
            aria-controls="1-content"
            aria-labelledby="1-title"
            index={0}
         >
            <h1 id="1-title">item one</h1>
            <AccordionIcon CloseIcon="ðŸ‘‡" OpenIcon="ðŸ‘ˆ" index={0} />
         </AccordionButton>
         <AccordionContent id="1-content" index={0}>
            <div>
               Horses can sleep both lying down and standing up. Domestic horses
               have a lifespan of around 25 years. A 19th century horse named
               'Old Billy' is said to have lived 62 years.
            </div>
         </AccordionContent>
      </AccordionContext.Provider>
   );
};
