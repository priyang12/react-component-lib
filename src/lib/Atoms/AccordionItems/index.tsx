export {
   AccordionButton,
   AccordionContent,
   AccordionIcon,
} from './AccordionItems';
