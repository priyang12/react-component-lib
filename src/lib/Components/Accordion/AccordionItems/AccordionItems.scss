.Accordion-content {
   overflow: hidden;
   padding: 1.5rem;
   background-color: var(--bg-default, #fff);
   color: var(--text-secondary, #444);
   font-size: 0.95rem;
   line-height: 1.5;
   transition: padding 0.3s ease;
}

.content-animation-enter {
   max-height: 0;
   opacity: 0;
   transform: translateY(-8px);
}

.content-animation-enter-active {
   max-height: 1000px;
   opacity: 1;
   transform: translateY(0);
   transition: max-height 0.4s ease, opacity 0.3s ease, transform 0.3s ease;
}

.content-animation-exit {
   max-height: 1000px;
   opacity: 1;
   transform: translateY(0);
}

.content-animation-exit-active {
   max-height: 0;
   opacity: 0;
   transform: translateY(-8px);
   transition: max-height 0.3s ease, opacity 0.2s ease, transform 0.3s ease;
}

.Accordion-icon {
   display: inline-flex;
   align-items: center;
   justify-content: center;
   transition: transform 0.3s ease;
   font-size: inherit;
   color: var(--text-secondary, #555);
}

.Accordion-icon--open {
   transform: rotate(180deg);
}
