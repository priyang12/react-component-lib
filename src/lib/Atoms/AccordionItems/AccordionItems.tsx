import * as React from 'react';
import { CSSTransition } from 'react-transition-group';
import { AccordionContext } from '../../Module/Accordion/Accordion';
import './AccordionItems.scss';

export interface AccordionButtonProps {
   index: number;
}

export function AccordionButton({
   index,
   children,
   ...props
}: AccordionButtonProps & React.ComponentPropsWithoutRef<'div'>) {
   const { handleItemClick, indexes } = React.useContext(AccordionContext);

   const ref = React.useRef<HTMLDivElement>(null);

   const KeyDown: React.ComponentProps<'div'>['onKeyDown'] = e => {
      const keyCode = e.code;
      if (keyCode === 'Enter' || keyCode === 'Space') handleItemClick(index);
      else if (keyCode === 'ArrowDown') {
         const nextItem = ref.current?.parentElement?.querySelector(
            `[data-index="${index + 1}"]`
         );
         if (nextItem) (nextItem as HTMLElement).focus();
      } else if (keyCode === 'ArrowUp') {
         const prevItem = ref.current?.parentElement?.querySelector(
            `[data-index="${index - 1}"]`
         );
         if (prevItem) (prevItem as HTMLElement).focus();
      }
   };

   return (
      <div
         className="Accordion-header"
         role="button"
         ref={ref}
         data-index={index}
         aria-expanded={indexes.includes(index)}
         tabIndex={0}
         onKeyDown={KeyDown}
         onClick={() => handleItemClick(index)}
         {...props}
      >
         {children}
      </div>
   );
}

interface AccordionContentType {
   index: number;
}

export function AccordionContent({
   index,
   children,
   ...props
}: AccordionContentType & React.ComponentPropsWithoutRef<'div'>) {
   const { indexes } = React.useContext(AccordionContext);
   return (
      <div {...props}>
         <CSSTransition
            in={indexes.includes(index)}
            timeout={200}
            classNames="content"
            mountOnEnter
            unmountOnExit
         >
            {children}
         </CSSTransition>
      </div>
   );
}

interface AccordionIconType {
   index: number;
   OpenIcon: React.ReactNode;
   CloseIcon: React.ReactNode;
}

export function AccordionIcon({
   index,
   CloseIcon,
   OpenIcon,
}: AccordionIconType & React.ComponentPropsWithoutRef<'span'>) {
   const { indexes } = React.useContext(AccordionContext);

   return <span>{indexes.includes(index) ? CloseIcon : OpenIcon}</span>;
}
