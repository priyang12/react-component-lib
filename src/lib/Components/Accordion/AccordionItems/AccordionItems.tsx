import * as React from 'react';
import { CSSTransition } from 'react-transition-group';
import { AccordionContext } from '../Accordion';
import './AccordionItems.scss';

export interface AccordionButtonProps {
   index: number;
}
/**
 * AccordionButton is a component that represents a button within an accordion.
 *
 * @param index - The index of the accordion item that the button belongs to.
 * @param children - The content of the button.
 * @param props - Additional props to be passed to the button element.
 *
 * @returns a div element that acts as a button.
 *
 * @example
 *
 * <AccordionButton index={0}>Click me</AccordionButton>
 */
export function AccordionButton({
   index,
   children,
   ...props
}: AccordionButtonProps & React.ComponentPropsWithoutRef<'div'>) {
   const { handleItemClick, indexes } = React.useContext(AccordionContext);

   const ref = React.useRef<HTMLDivElement>(null);

   const KeyDown: React.ComponentProps<'div'>['onKeyDown'] = (e) => {
      const keyCode = e.code;
      if (keyCode === 'Enter' || keyCode === 'Space') handleItemClick(index);
      else if (keyCode === 'ArrowDown') {
         const nextItem = ref.current?.parentElement?.querySelector(
            `[data-index="${index + 1}"]`
         );
         if (nextItem) (nextItem as HTMLElement).focus();
      } else if (keyCode === 'ArrowUp') {
         const prevItem = ref.current?.parentElement?.querySelector(
            `[data-index="${index - 1}"]`
         );
         if (prevItem) (prevItem as HTMLElement).focus();
      }
   };

   return (
      <div
         className="Accordion-header"
         role="button"
         ref={ref}
         data-index={index}
         aria-expanded={indexes.includes(index)}
         tabIndex={0}
         onKeyDown={KeyDown}
         onClick={() => handleItemClick(index)}
         {...props}
      >
         {children}
      </div>
   );
}

interface AccordionContentType {
   index: number;
}
/**
 * AccordionContent is a component that represents the content of an accordion item.
 *
 * @param index - The index of the accordion item that the content belongs to.
 * @param children - The content of the accordion item.
 * @param props - Additional props to be passed to the div element.
 *
 * @returns a div element that contains the accordion item content.
 *
 * @example
 *
 * <AccordionContent index={0}>Item content</AccordionContent>
 */
export function AccordionContent({
   index,
   children,
   ...props
}: AccordionContentType & React.ComponentPropsWithoutRef<'div'>) {
   const { indexes } = React.useContext(AccordionContext);
   return (
      <div {...props}>
         <CSSTransition
            in={indexes.includes(index)}
            timeout={200}
            classNames="content"
            mountOnEnter
            unmountOnExit
         >
            {children}
         </CSSTransition>
      </div>
   );
}

interface AccordionIconType {
   index: number;
   OpenIcon: React.ReactNode;
   CloseIcon: React.ReactNode;
}
/**
 * AccordionIcon is a component that represents an icon within an accordion item.
 *
 * @param index - The index of the accordion item that the icon belongs to.
 * @param OpenIcon - The icon to be displayed when the accordion item is open.
 * @param CloseIcon - The icon to be displayed when the accordion item is closed.
 *
 * @returns a span element that contains the appropriate icon.
 *
 * @example
 *
 * <AccordionIcon
 *   index={0}
 *   OpenIcon={<i className="fas fa-chevron-up" />}
 *   CloseIcon={<i className="fas fa-chevron-down" />}
 * />
 */
export function AccordionIcon({
   index,
   CloseIcon,
   OpenIcon,
}: AccordionIconType & React.ComponentPropsWithoutRef<'span'>) {
   const { indexes } = React.useContext(AccordionContext);

   return <span>{indexes.includes(index) ? CloseIcon : OpenIcon}</span>;
}
