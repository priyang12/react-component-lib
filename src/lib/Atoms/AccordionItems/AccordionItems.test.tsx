import { render, fireEvent, screen } from '@testing-library/react';
import { composeStories } from '@storybook/testing-react';
import * as AccordionItemsStories from './AccordionItems.stories';

const { Template } = composeStories(AccordionItemsStories);

it('renders correctly', () => {
   const { container } = render(<Template />);
   expect(container).toMatchSnapshot();
});
it('handles key press events correctly', () => {
   const handleKeyDown = jest.fn();
   render(<Template onKeyDown={handleKeyDown} />);
   fireEvent.keyDown(screen.getByText('item one'), { code: 'Enter' });
   expect(handleKeyDown).toHaveBeenCalled();
});
