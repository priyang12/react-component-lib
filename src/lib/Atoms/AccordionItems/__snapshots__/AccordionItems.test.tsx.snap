// Vitest Snapshot v1

exports[`renders correctly 1`] = `
<div>
  <div
    class="container"
  >
    <div
      aria-controls="1-content"
      aria-expanded="false"
      aria-labelledby="1-title"
      class="Accordion-header"
      data-index="0"
      role="button"
      tabindex="0"
    >
      <h1
        id="1-title"
      >
        item one
      </h1>
      <span>
        ðŸ‘ˆ
      </span>
    </div>
    <div
      id="1-content"
    />
  </div>
</div>
`;
